python/pyarrow/parquet/core.py

```diff
@@ -3423,15 +3427,17 @@ def write_metadata(schema, where, metadata_collector=None, **kwargs):
     ...     table.schema, 'dataset_metadata/_metadata',
     ...     metadata_collector=metadata_collector)
     """
-    writer = ParquetWriter(where, schema, **kwargs)
+    writer = ParquetWriter(where, schema, filesystem, **kwargs)
     writer.close()
 
     if metadata_collector is not None:
         # ParquetWriter doesn't expose the metadata until it's written. Write
         # it and read it again.
-        metadata = read_metadata(where)
+        metadata = read_metadata(where, filesystem=filesystem)
         for m in metadata_collector:
             metadata.append_row_groups(m)
+        if filesystem is not None:
+            filesystem, where = _resolve_filesystem_and_path(where, filesystem)
```

Comment:
Shouldn't this go up higher, if you want to support inferring the filesystem from the path? (Even if those other function also do the inference, it seems like it would be nice if the inference was only performed once.